docs/architecture.md

```diff
@@ -8,16 +8,109 @@ This whole application was built off of [Node.js](https://nodejs.org/en/)!
+For authorization, we decided to write a custom RBAC-esque strategy. Check below for specifics on how it works, and how to work with it, and how to use it when extending this project.
 
 ### Testing
 
 We use [Mocha](https://mochajs.org/) with [Chai](https://www.chaijs.com/) to test our routes and services.
 
+## Authorization
+
+Before you read this section, it's important to understand the difference between `Authentication` and `Authorization`. The first asks the question: _"Do we know who you are?"_ (i.e. are you logged in?), whereas the second asks the question: _"Do you have sufficient permissions to be making this request?"_.
+
+We use a custom implementation for authorization, which relies heavily on the design of Role-based access control.
+
+### Definitions
+
+#### `Route`
+
+A `Route` is defined by its uri path, its query parameters in the uri path, and the HTTP verb used to access it:
+
+```json
+{
+    "requestType": "POST",
+    "uri": "/api/sponsor/",
+}
+```
+
+#### `Role`
+
+A `Role` is a collection of `Routes`, and a unique name (such as `hacker`, or `sponsor`, or `my-custom-role`). This collection of `Routes` should approximately encapsulate the set of actions that a user of the api should be able to do. For example, a `Hacker` might want to have access to `GET /api/hacker/507f1f77bcf86cd799439011` (get their own information), or `POST /api/hacker/` (create their hacker document):
+
+```json
+{
+    "accountId": "507f191e810c19729de860ea",
+    "roles": ["54759eb3c090d83494e2d804", "51325eb3c090d83494e2d702"],
+}
+```
+
+_note here that `roles` is a list of `ObjectIds`, since the role is also a document in the database._
+
+#### `RoleBinding`
+
+A `RoleBinding` is a mapping between an `account` and a `Role`. In this way, we can say that an account has a certain set of allowed actions, defined by the `Role`. An `account` can have a `RoleBinding` to multiple `Roles`.
+
+### How to write a `Route` document
+
+In order to encapsulate permissions based on the actual `params` of a given route (such as the resource id in `/api/hacker/507f1f77bcf86cd799439011`), we use two placeholders: `:self`, and `:all`. These two placeholders replace the location of the given parameter in the route; in the example above, we would translate the uri to either `/api/hacker/:self`, or `/api/hacker/:all`.
+
+#### `:self`
+
+The `self` placeholder provides the account that hits this `Route` access to only the resource that is related to this particular accountId. For example, if I wanted to hit `GET /api/hacker/507f1f77bcf86cd799439011`, then the hacker whose `id` was `507f1f77bcf86cd799439011` would have to have an `accountId` that is equal to the `id` of the account accessing this resource.
+
+#### `:all`
+
+The `all` placeholder provides the account that hits this `Route` access to any resource for a given id.
+
+#### Chaining :self and :all
+
+It is possible to chain `:self` and `:all` together if need-be.
```

Comment:
What do you mean by this?
Does chaining mean /api/hacker/:self:all, or /api/hacker/:self/:all?
We do allow for the second option, but the first type is not handled.

Reply:
The second case